### PR TITLE
github action implemented for the HW2

### PR DESCRIPTION
**Pull request by tirthbha**
github action is  implemented for the HW2 i.e for the matrix multiplication using three different approaches: For Loops, NumPy Built-in Functions, and NumPy Einstein Notation Function.

### CODE REVIEW

File: .github/workflows/matrix_multiplication_unit_test.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        echo "numpy==1.21.0" > requirements.txt
+        echo "matplotlib==3.4.2" >> requirements.txt
+        pip install -r requirements.txt
+    - name: List directory contents  # Debugging step to see contents
+      run: ls -al
+    - name: Run matrix multiplication tests, generate plot, and write to log
+      run: python test_matrix_multiplication.py
+
+    - name: Upload plot and log as artifacts
+      uses: actions/upload-artifact@v2
```
Comment (laserlab) [inline .github/workflows/matrix_multiplication_unit_test.yml:32]: That is a nice idea!

File: .github/workflows/Unittesting.yml [hunk 2]
```diff
@@ -6,16 +6,14 @@ jobs: # what should we do when it is run
...
         python-version: '3.10' # specify python version
     - name: Install dependencies # which libraries do we need for testing
       run: |
         python -m pip install --upgrade pip
         pip install -r requirements.txt # they should be listed in this file
     - name: Test with pytest
-      run: |
-        pip install pytest
-        pytest tests.py
+      run: pytest test_matrix_multiplication.py
```
Comment (laserlab) [inline .github/workflows/Unittesting.yml:19]: You’re added line will work, but removing the run statement likely will make it fail

File: .github/workflows/Unittesting.yml [hunk 3]
```diff
@@ -6,16 +6,14 @@ jobs: # what should we do when it is run
   build:
     runs-on: ubuntu-latest # platform it runs on
     steps:
-    - uses: actions/checkout@v4 # checkout your code
+    - uses: actions/checkout@v2 # checkout your code
```
Comment (laserlab) [inline .github/workflows/Unittesting.yml:9]: Does this make any difference?

File: .github/workflows/python_pytest_lint.yml. [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    - name: Lint with flake8
+      run: |
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=76 --statistics
+
+    - name: Run unit tests with pytest
+      run: |
+        pip install pytest
+        pytest matrix_multiplication.py
```
Comment (laserlab) [inline .github/workflows/python_pytest_lint.yml.:29]: I would keep that in the seperate action

**Review by laserlab — CHANGES_REQUESTED** [review]
Looks good.

Be aware you don’t have the ‘ test_matrix_multiplication.py’ file

**Comment by tirthbha** [comment]
> Looks good.
> 
> Be aware you don’t have the ‘ test_matrix_multiplication.py’ file

Actually, I have changed the file name. The file is in the directory.

File: .github/workflows/Unittesting.yml (hunk 2 above)
Comment (tirthbha) [reply .github/workflows/Unittesting.yml:19]: Yeah professor, I tried this.

File: .github/workflows/Unittesting.yml (hunk 3 above)
Comment (tirthbha) [reply .github/workflows/Unittesting.yml:9]: Actually, Initially I got error with the later one so I added the former one, and it worked. I will check this again.

**Pull request merged**